### PR TITLE
[spirv] Add VK_KHR_variable_rate_fragment_shading SPIR-V support

### PR DESCRIPTION
**Pull request by LLJJDD**
Change the mapping of SV_ShadingRate from FragSizeEXT (VK_EXT_fragment_density_map) to PrimitiveShadingRateKHR/ShadingRateKHR (VK_KHR_variable_rate_fragment_shading).

### CODE REVIEW

**Comment by jaebaek** [comment]
@LLJJDD Thank you for sending this PR. I will take a look.

**Comment by jaebaek** [comment]
I am confused .. based on the [SPV_KHR_fragment_shading_rate spec](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_fragment_shading_rate.html#_modifications_to_the_spir_v_specification_version_1_5), it seems to be added from SPIR-V 1.5. Wasn't it?

If it is correct, we have to keep the previous behavior for SPIR-V 1.0 to 1.4 and apply your change only for SPIR-V 1.5.

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 1]
```diff
@@ -3470,19 +3446,16 @@ SpirvVariable *DeclResultIdMapper::createSpirvStageVar(
...
-  // According to DXIL spec, the ShadingRate SV can only be used by GSOut,
-  // VSOut, or PSIn. According to Vulkan spec, the FragSizeEXT BuiltIn can only
-  // be used as PSIn.
   case hlsl::Semantic::Kind::ShadingRate: {
     switch (sigPointKind) {
     case hlsl::SigPoint::Kind::PSIn:
       stageVar->setIsSpirvBuiltin();
-      return spvBuilder.addStageBuiltinVar(type, sc, BuiltIn::FragSizeEXT,
+      return spvBuilder.addStageBuiltinVar(type, sc, BuiltIn::ShadingRateKHR,
                                            isPrecise, srcLoc);
```
Comment (jaebaek) [inline tools/clang/lib/SPIRV/DeclResultIdMapper.cpp:3454]: We have to keep the error report for the default case, but want to slightly change it e.g.,

```
    default:
      emitError("semantic ShadingRate must be used only for PSIn, VSOut, GSOut,"
                "  MSOut",
                srcLoc);
      break;
```

In addition, we have to add the following cases:

```
case hlsl::SigPoint::Kind::VSOut:
case hlsl::SigPoint::Kind::GSOut:
case hlsl::SigPoint::Kind::MSOut:
      stageVar->setIsSpirvBuiltin();
      return spvBuilder.addStageBuiltinVar(
          type, sc, BuiltIn::PrimitiveShadingRateKHR, isPrecise, srcLoc);
```

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 2]
```diff
@@ -3470,19 +3446,16 @@ SpirvVariable *DeclResultIdMapper::createSpirvStageVar(
     return spvBuilder.addStageBuiltinVar(type, sc, BuiltIn::FullyCoveredEXT,
                                          isPrecise, srcLoc);
   }
-  // According to DXIL spec, the ShadingRate SV can only be used by GSOut,
-  // VSOut, or PSIn. According to Vulkan spec, the FragSizeEXT BuiltIn can only
-  // be used as PSIn.
   case hlsl::Semantic::Kind::ShadingRate: {
     switch (sigPointKind) {
```
Comment (jaebaek) [inline tools/clang/lib/SPIRV/DeclResultIdMapper.cpp:3450]: It seems `SV_ShadingRate` is only allowed to be used for `uint` type.
(See https://github.com/KhronosGroup/SPIRV-Tools/blob/master/test/val/val_builtins_test.cpp#L3842)

I asked this question to doublecheck #3676 .

Why don't we add type check here?

e.g.,
```
case hlsl::Semantic::Kind::ShadingRate: {
QualType ty;
if (!isScalarType(type, &ty) || !ty->isIntegerType()) {
// report error
}
```

`bool isScalarType(QualType type, QualType *scalarType)` is defined in tools/clang/lib/SPIRV/AstTypeProbe.cpp file.

**Comment by LLJJDD** [comment]
> I am confused .. based on the [SPV_KHR_fragment_shading_rate spec](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_fragment_shading_rate.html#_modifications_to_the_spir_v_specification_version_1_5), it seems to be added from SPIR-V 1.5. Wasn't it?
> 
> If it is correct, we have to keep the previous behavior for SPIR-V 1.0 to 1.4 and apply your change only for SPIR-V 1.5.

The spec is `written against the SPIR-V Specification, Version 1.5, Revision 3, Unified`, but it only `requires SPIR-V 1.0`.

It works fine with SPIR-V 1.0 to 1.5.

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp (hunk 2 above)
Comment (LLJJDD) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp:3450]: There will be an implicit float->uint type conversion in such case. Anyway, adding a type check won't hurt.

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 3]
```diff
@@ -3472,18 +3448,32 @@ SpirvVariable *DeclResultIdMapper::createSpirvStageVar(
...
-      return spvBuilder.addStageBuiltinVar(type, sc, BuiltIn::FragSizeEXT,
+      return spvBuilder.addStageBuiltinVar(type, sc, BuiltIn::ShadingRateKHR,
                                            isPrecise, srcLoc);
+    case hlsl::SigPoint::Kind::VSOut:
+    case hlsl::SigPoint::Kind::GSOut:
+    case hlsl::SigPoint::Kind::MSOut:
+      stageVar->setIsSpirvBuiltin();
+      return spvBuilder.addStageBuiltinVar(
+          type, sc, BuiltIn::PrimitiveShadingRateKHR, isPrecise, srcLoc);
+      break;
```
Comment (jaebaek) [inline tools/clang/lib/SPIRV/DeclResultIdMapper.cpp:3472]: remove this `break;`

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 4]
```diff
@@ -3472,18 +3448,32 @@ SpirvVariable *DeclResultIdMapper::createSpirvStageVar(
...
     default:
-      emitError("semantic ShadingRate currently unsupported in non-PS shader"
-                " stages",
+      emitError("semantic ShadingRate must be used only for PSIn, VSOut, "
+                "GSOut, MSOut",
                 srcLoc);
-      break;
```
Comment (jaebaek) [inline tools/clang/lib/SPIRV/DeclResultIdMapper.cpp:3486]: do not remove this `break;`

**Review by jaebaek — COMMENTED** [review]
LGTM with nits.

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp (hunk 4 above)
Comment (LLJJDD) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp:3486]: Oops.

**Review by jaebaek — APPROVED** [review]

**Comment by jaebaek** [comment]
Thank you for your contribution!

**Pull request merged**